### ISSUE TITLE
WebGPU support

### ISSUE THREAD

**Issue by sluijs** [huggingface/candle#344]
Is WebGPU support on the roadmap as an alternative GPU-accelerated backend? This would be especially useful for inference on the web or for non-CUDA environments.

**Comment by LaurentMazare** [comment]
WebGPU is certainly on our radar, we already have some wasm based demos for [llama2.c](https://huggingface.co/spaces/lmz/candle-llama2) and [whisper](https://huggingface.co/spaces/lmz/candle-whisper) that you can try in a web browser. When using wasm, candle should leverage your cpu simd instructions, but having WebGPU on top of this would bring it to a far better level.

**Comment by ethanhs** [comment]
Hi! I'd be interested in working on this. I've spent some time thinking about a rough plan after reading through the code:

1. Move `candle-kernels` to `candle-cuda` or `candle-cuda-kernels` (the name can be bikeshed'd in the PR)
2. Make a `candle-wgsl(-kernels)` crate, with kernels implementing the ops needed. Can maybe re-use some implementations based on https://github.com/webonnx/wonnx
3. Add a new backend implementation using wgpu-rs to execute the kernels
4. Add tests, info on how to run things
5. Maybe add flash attention kernels -- might be a lot of work so probably worth its own follow-up issue.

Some questions:
- Is wrapping with [wgpu-rs](https://github.com/gfx-rs/wgpu) acceptable? I see that the cuda wrappers use cudarc, and wgpu-rs seems to be the closest equivalent for wgsl shaders for compute. It would also be an easy way to bridge things on native to Vulkan and Metal as well (would take more work though, e.g. integrating [naga](https://github.com/gfx-rs/naga)).
- Do you want all kernels in one crate? In the above I suggest splitting them into their own backend-specific crates, but I guess that could be abstracted into the `candle-kernels` crate itself if you prefer.
- I will probably focus on the web for now, because I think it is the most promising, but I also will probably follow up with testing Vulkan, as I think that will make inference much more portable.

**Comment by Narsil** [comment]
Sounds like a  very reasonable plan.

I think we can start working without tying too much to `candle`, maybe other projects could be interested on having webgpu support (that's why having cudarc is great it can be used by other projects, not necessarily candle, and why we keep pushing changes upstream as much as possible, like the NCCL support).

`wgpu-rs` : Last time I tried Vulkan, and doing compute shaders, the performance was abysmal. And it makes sense, it's not really designed for ML. In general I would go for the most performant solutions from the start, not have backends just for the sake of it. AMD already has libraries intended for ML: https://www.amd.com/en/graphics/servers-solutions-rocm we could link to that directly if it makes more sense.

For Metal I would have the same opinion, we should try and make metal usable outside of this crate and be mere users of it.

For any new backend, it is very important to create a way for USERS to create their own kernel/op. It's impossible to keep up with all the innovation imho so the most important thing is to allow users of candle to use any op they want, without having to wait for us to implement it.

**Comment by ethanhs** [comment]
re wgpu-rs, I certainly agree that native backends are the best, I only bring up Vulkan/Metal as bonuses. I was suggesting wgpu-rs because it is the major WebGPU library for Rust, it looks like [Burn uses it](https://github.com/burn-rs/burn/tree/main/burn-wgpu). So I think it is the best library for the job, I just wanted to see if adding the dependency was acceptable. The alternative would be to write a bunch of bindings via web-sys around webgpu APIs.

> For any new backend, it is very important to create a way for USERS to create their own kernel/op. 

Certainly! I mostly was discussing the crate rename/split focused on candle-provided kernels. For user written kernels, would it not be best to simply add `wgpu_fwd` to the `Op{N}` traits that the user may implement? Are there other details I should be aware of?

**Comment by Narsil** [comment]
> Certainly! I mostly was discussing the crate rename/split focused on candle-provided kernels. For user written kernels, would it not be best to simply add wgpu_fwd to the Op{N} traits that the user may implement? Are there other details I should be aware of?

Basically yes. Tensor is Send+Sync, therefore `Op` needs to be `Send+Sync` (because it's kept for gradients). That could end up being a limitation: https://github.com/huggingface/candle/blob/main/candle-examples/examples/llama_multiprocess/model.rs#L33-L38

> I think it is the best library for the job

What other libraries or alternatives are there ?
Looking at this: https://www.reddit.com/r/rust/comments/159cbto/announcing_burnwgpu_new_deep_learning/jtf80xq/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button I have the feeling like it's not the correct way. We need only webGPU, not those 10 other things.
In any case this is not in our short term roadmap.

**Comment by burke-up** [comment]
is now support webgpu ?

**Comment by guyoung** [comment]
candle webassembly Is there any plan to support WebGPU？

**Comment by minghuaw** [comment]
One general comment.

> Move candle-kernels to candle-cuda or candle-cuda-kernels (the name can be bikeshed'd in the PR)

I feel like writing those compute shaders in glsl might be a better option. I have done some rough testing on different gpgpu performance and vulkan with glsl seems to be able to keep up with cuda while webgpu with wgsl reaches bottleneck pretty early with the same optimization tricks. On top of that, webgpu supports glsl as well, so we could have not only a webgpu backend but a vulkan one as well (I guess for folks who still want to run it natively by don't have the luxury of a nvidia GPU but an intel/AMD GPU)

**Comment by santiagomed** [comment]
@LaurentMazare are there any plans to start implementing a WebGPU backend? I see [Ratchet](https://github.com/huggingface/ratchet) has successfully implemented WebGPU inference and would love to see this in Candle soon as well. I would love to help with this implementation if possible too, but there's a lot of learning on my part to be done before I do so, so ideally would love to chat more since I'd have a lot of questions.

**Comment by cgisky1980** [comment]
https://github.com/cryscan/web-rwkv 
Here is a RWKV LLM inference based on WGPU  Vulkan。

**Comment by KimHenrikOtte** [comment]
Maybe this fork will help if WebGpu support is requested: 
As part of a university project this semester, I developed a candle fork with a WGPU backend. (https://github.com/KimHenrikOtte/candle/tree/wgpu_test)
I created implementations for most of the internal operations. Currently only sorting and a few operations in candle-nn are not implemented. 

In the end, I managed to create an image in the browser using a newly created wuerstchen wasm example. (I don't think this would be possible otherwise as there is a 4GB ram limit and the model is too big to be held by webassembly). 
 
 I used wgpu as the backend. So the implementations should also work native with Vulkan, Metal or DirectX.
 All kernels are implemented by myself.
 I tried my best to make the shaders more or less performant. But for some shaders (e.g. matmul) it is quite hard to get the best performance in all scenarios. 
 
# Feature Support Table

| Feature                     | Support Status                                  | Notes                                                              |
|-----------------------------|-------------------------------------------------|--------------------------------------------------------------------|
| **<span style="color:#1E90FF">Data Types</span>**                                  |                                       |                                                                    |
| f32                         | ✅ Supported                                     |                                                                    |
| u32                         | ✅ Supported                                     |                                                                    |
| u8                          | ⚠️ Only Output of Cmp                            | *Only f32, I32 and U32 are available in a webGpu shader            |
| i64                         | ❌ Not Implemented                               |                                                                    |
| f64                         | ❌ Not Supported                                 |                                                                    |
| f16                         | ❌ Not Supported                                 |                                                                    |
| bf16                        | ❌ Not Supported                                 |                                                                    |
| **<span style="color:#1E90FF">Operations</span>**              |                                                 |   All operations support non-contiguous arrays                                                                   |
| Unary Operations            | ✅ Supported                                     |                                                   |
| Binary Operations           | ✅ Supported                                     |                                |
| MatMul                      | ✅ Supported                                     |                                                                    |
| Reduce Operations           | ✅ Supported                                     | Sum, Min, Max, (ArgMax, ArgMin works only if continues Dimensions are reduced)                                     |
| Conv2d                      | ✅ Supported                                     |                                                                    |
| Conv2dTranspose             | ✅ Supported                                     |       |
| Conv1d                      | ✅ Supported                                     |                                                                    |
| Conv1dTranspose             | ✅ Supported                                     |                                                                    |
| Index Select                | ✅ Supported                                     |                                                                    |
| Where_cond                  | ✅ Supported                                     |                                                                    |
| Pool2dMax                   | ✅ Supported                               |                                                                    |
| Pool2dAvg                   | ✅ Supported                               |                                                                    |
| Upsample                    | ✅ Supported                               |                                                                    |
| Gather                      | ✅ Supported                               |                                                                    |
| Scatter_add                 | ✅ Supported                               |                                                                    |
| Index_add                   | ✅ Supported                              |                                                                    |
| **<span style="color:#1E90FF">Not Implemented</span>**        |                                                 |                                                                    |
| ArgSort                     | ❌ Not Implemented                               |                                                                    |
| Quantized Matrices          | ❌ Not Supported?                                 |                                                                    |

**Comment by sluijs** [comment]
@LaurentMazare any chance that the proposed backend/webgpu kernels can become integrated at some point?

**Comment by sinkingsugar** [comment]
That would be a big (and maybe easy thanks to @KimHenrikOtte work) win for candle which already builds so easily on every platform and device (our app ships on android, vision, ios etc and works well with candle) 👍

**Comment by EdupugantiAkhil** [comment]
@KimHenrikOtte I am interested in contributing to your project if you are willing to accept PR's to your repo. I ended up playing around with it to get clip working with slight(draft) modifications [here](https://github.com/EdupugantiAkhil/candle_webgpu).

**Comment by KimHenrikOtte** [comment]
I did some more work on the branch. 
Currently the latest version is at: https://github.com/KimHenrikOtte/candle/tree/wgpu_cleanup
Current state of the branch:
- Wgpu backend added
- Support for f32, u32, u8(Cmp and WhereCond), i64(native) and f64(native)
- Support for all candle core operations, except asort
- Support for softmax, rms-norm, layer-norm, sigmoid 
- New  `to_vecX_async`,  `to_scalar_async` and  `to_device_async` functions added for use in the browser.
- All candle-wasm-examples support the wgpu async implementation.
- You can define your own WGSL shaders.
- Converted all tests to WASM-compatible tests with a custom build.rs
   -> All tests pass
- Added  `wgpu_basics.rs` example to demonstrate how to use custom shaders.
- Documented how to use the wgpu backend in the candle book.
- all new dependencies are behind the `wgpu` feature
- removed all debug/test files
- fixed most clippy warnings

Current issues: 
- There are 3 clippy warnings about a `MutexGuard` being held over an `await` point. 
  I am not sure how to fix this, as the mutex needs to be accessed within sync and async functions.
  I think as long as you are not awaiting on multiple candle functions in parallel, this is not a problem. 
- Implementation/performance tested on Windows with NVIDIA GPU only.

**Comment by mrchantey** [comment]
I was just thinking about this feature this morning so happy to hear its bring worked on, being able to target WebGPU will unlock so many cool web demos 🤩

**Comment by KimHenrikOtte** [comment]
I want to draw more attention to the WGPU fork: [https://github.com/KimHenrikOtte/candle/tree/wgpu_cleanup](https://github.com/KimHenrikOtte/candle/tree/wgpu_cleanup)

I do not have the resources to test all possible examples or hardware configurations, so I cannot guarantee that everything works perfectly across all scenarios. However, all changes are gated behind the `wgpu` feature flag, meaning the default functionality should remain unaffected for users who do not enable it. For those who enable the feature flag, this update provides an opportunity to experiment with and test the wgpu backend. 

Do you think this branch is in a state where I could open a PR?

## Performance Analysis
Although I couldn’t get CUDA running on my system for a proper comparison, the WGPU backend performs significantly better than the CPU implementation for larger models on my test hardware:

**Test Hardware**:  
- **CPU**: AMD Ryzen 7 5800X (8-Core Processor)  
- **GPU**: NVIDIA GeForce GTX 1080 Ti  
- **RAM**: 3200 MHz  


| **Model/Application**  | **Native (WGPU)** | **Native (CPU)** | **Browser (WGPU)** | **Browser (CPU)** |
|-------------------------|-------------------|------------------|--------------------|-------------------|
| **Llama2-c 15m**        | 289.20 tok/s     | 359.70 tok/s     | 117.49 tok/s       | 85.72 tok/s       |
| **Llama2-c 42m**        | 209.09 tok/s     | 65.62 tok/s      | 109.76 tok/s       | 47.80 tok/s       |
| **T5**                 | 102.21 tok/s     | 77 tok/s         |                  |                |
| **Clip**               | 1.15 s           | 947.92 ms        |                  |               |
| **Metavoice**          | 59.50 s          | 617.39 s         |                 |               |
| **Stable Diffusion**   | 27.62 s          | 338.66 s         | 65 s            | N/A               |
| **Wuerstchen**         | 131.30 s         | >1740.00 s        | 160 s                | N/A               |

Commands used to test the models:
```bash
cargo run --example llama2-c --release --features "candle-datasets,wgpu" -- inference --which-model stories15M.bin 
cargo run --example llama2-c --release --features "candle-datasets,wgpu" -- inference --which-model stories42M.bin
cargo run --example t5 --release --features="wgpu" -- --model-id "t5-small" --prompt "translate to German: A beautiful candle is glowing at a christmas table. It is very cold outside and snowing." --decode
cargo run --example clip --release --features="wgpu" -- --images "examples/stable-diffusion/assets/stable-diffusion-xl.jpg,examples/yolo-v8/assets/bike.jpg" --sequences  "a cycling race,a photo of two cats,a robot holding a candle"
cargo run --example metavoice --release --features=wgpu -- --prompt "This is a demo of text to speech by MetaVoice-1B, an open-source foundational audio model." --out-file "out_meta_wgpu.wav"
cargo run --example stable-diffusion --release --features=wgpu  -- --prompt "Anthropomorphic cat dressed as a fire fighter" --sd-version v1-5 --final-image "out_sd_wgpu.png"
cargo run --example wuerstchen --release  --features="wgpu" -- --prompt "realistic cat dressed as a fire fighter" --final-image "out_wu_wgpu.png"



cargo run --example llama2-c --release --features "candle-datasets,wgpu" -- --cpu inference --which-model stories15M.bin 
cargo run --example llama2-c --release --features "candle-datasets,wgpu" -- --cpu inference --which-model stories42M.bin
cargo run --example t5 --release --features="wgpu" -- --model-id "t5-small" --prompt "translate to German: A beautiful candle is glowing at a christmas table. It is very cold outside and snowing." --decode --cpu
cargo run --example clip --release --features="wgpu" -- --images "examples/stable-diffusion/assets/stable-diffusion-xl.jpg,examples/yolo-v8/assets/bike.jpg" --sequences  "a cycling race,a photo of two cats,a robot holding a candle"  --cpu
cargo run --example metavoice --release --features=wgpu -- --prompt "This is a demo of text to speech by MetaVoice-1B, an open-source foundational audio model." --out-file "out_meta_cpu.wav"  --cpu
cargo run --example stable-diffusion --release --features=wgpu  -- --prompt "Anthropomorphic cat dressed as a fire fighter" --sd-version v1-5 --final-image "out_sd_cpu.png"  --cpu
cargo run --example wuerstchen --release  --features="wgpu" -- --prompt "realistic cat dressed as a fire fighter" --final-image "out_wu_cpu.png"  --cpu
```

**Comment by Quackdoc** [comment]
I was able to test whisper-microphone on my A380 and it seems to be working on. I did try stable diffusion but predictably I wasn't able to generate anything, and when I tried passing --{width,height} 256 it managed to get all the way through until the last step where it locked up. 192 and below "worked" in so much that they generated an image, but no prompt made any thing that resembled the prompt. Maybe f16 support in the future might help it be more usable? metavoice just caused my GPU to run at 16mhz.

**Comment by KimHenrikOtte** [comment]
> I was able to test whisper-microphone on my A380 and it seems to be working on. I did try stable diffusion but predictably I wasn't able to generate anything, and when I tried passing --{width,height} 256 it managed to get all the way through until the last step where it locked up. 192 and below "worked" in so much that they generated an image, but no prompt made any thing that resembled the prompt. Maybe f16 support in the future might help it be more usable? metavoice just caused my GPU to run at 16mhz.

I suspect that Stable Diffusion’s performance degrades significantly at smaller image sizes. Even on the CPU, using dimensions of 256 seems to produce images that fail to align with the prompts. This seems to be a limitation of the model itself, rather than an issue with the backend.

As for `wgpu`, it currently doesn’t support f16, as outlined in this [GitHub issue](https://github.com/raphamorim/rio/issues/758). However, it appears that work is already underway to address this.

**Comment by mrchantey** [comment]
> Do you think this branch is in a state where I could open a PR?

I think opening a pr would be a great next step, it will provide more clarity on the proposal and create a space where others can review and contribute, even if its not yet merge ready 😸

**Comment by Quackdoc** [comment]
@KimHenrikOtte wgpu f16  support has been merged!

**Comment by theoparis** [comment]
I tried to use the wgpu_cleanup branch but it tries to build cudarc, which doesn't work on a ARM macbook machine...

**Comment by DrJesseGlass** [comment]
@LaurentMazare @Narsil 

I'm very interested in using WebGPU with Candle for web-based inference and would like to contribute to this effort.
@KimHenrikOtte has a nearly complete implementation that's been validated by multiple community members. 

What's blocking integration of this work into main? Getting this merged sooner (even if marked as experimental) would allow the community to contribute improvements incrementally. Is there a specific technical concern or roadmap issue preventing this from moving forward?

**Comment by ivarflakstad** [comment]
Hi everyone!

TLDR: Seems silly to not integrate nearly complete implementation. @KimHenrikOtte feel free to open some PRs if you'd like :) (I'd prefer to do it in chunks and from the original author, if possible. Easier to review and of course less chance of malicious code slipping in).

Long:
I've been looking at WebGPU integration for a while, ref ~100 commits in [ratchet](https://github.com/huggingface/ratchet) (shoutout to @FL33TW00D webgpu 🐐 ).

The problem is that candle is not designed with many backends in mind. The variants in the `Storage` and `Device` enums increases linearly with every backend added and quickly becomes hard to maintain.

To circumvent this I've begun work on a large refactoring in the [v2](https://github.com/huggingface/candle/tree/v2) branch. Instead of enums the tensor would be `Tensor<B: Backend>`. It works. It's functional. But there are some issues.
For example rust lacks specialization, or an ability to fall back to default behaviour if a trait is not implemented. So I started working on adding that capibility to rust itself [here](https://github.com/rust-lang/rust/pull/144363). This would be very useful for candle regardless of the refactoring going through or not, and obviously for a lot of other projects written in rust, but I haven't had the time to put the finishing touches on it.

So my next plan was to add a new backend (in the original enum based scheme) simply called `Lazy`. Tracks all operations, builds a graph, partial optimizations, all the standard stuff that everyone is doing these days, and then you (finish optimizations and) execute it via something that implements `LazyBackend` (name pending).
This means that instead of being dependant on the enums, you simply have to implement `LazyBackend`. Completely circumventing the need for the ever-growing enums.

But with all that said I refer back to the TLDR at the top :)

**Comment by KimHenrikOtte** [comment]
Hi everyone

Thanks a lot for the interest and discussion around adding **WebGPU support for Candle** — it’s great to see the community’s enthusiasm here!

Just to give a quick update on why I temporarily paused work on the WebGPU backend:

* I had to focus on another project for a while, so progress was put on hold.
* I’m still a bit unsatisfied with the **matrix multiplication implementation selection**. The backend doesn’t always pick the best-performing variant automatically, and finding the optimal implementation across different devices and GPU drivers is a really hard problem. For example, I’ve seen cases where an implementation that performs best in a native Windows application runs **up to 100× slower** in the browser, with very limited debugging options for the generated shader.
* I couldn’t **test on other hardware**, such as macOS/Metal systems, GPUs with f16 shader support, or larger neural networks, due to hardware limitations.

There are a few **minor additions to the main Candle API**, such as:

* `device.is_dtype_available()` – to check at runtime if a certain dtype (like f16) is supported.
* Async GPU readback methods like `synchronize_async`, `to_scalar_async`, `to_vec*_async`, and `to_device_async`.

Most of the WebGPU backend implementation itself is **guarded behind the `wgpu` feature flag**, meaning it’s only compiled and active when explicitly enabled. Even if the current state could still be considered somewhat experimental, this setup makes it safe to include — users have to actively enable the feature, and can simply disable it again if it causes any issues. This ensures that the WebGPU code doesn’t interfere with or affect other backends like CPU.

In the coming days, I plan to spend some time **cleaning up the repository** — for example, removing unnecessary whitespace changes, reviewing comments, and doing some general code tidying — to make it easier to review and hopefully move toward integration.

**Comment by ErichDonGubler** [comment]
Hey Kim, wgpu/Firefox maintainer here. If you need any help getting integrated with the WebGPU ([Matrix link](https://matrix.to/#/#WebGPU:matrix.org)) or wgpu ([Matrix link](https://matrix.to/#/#wgpu:matrix.org)) communities for questions, feel free to contact us!

**Comment by vandana-rajan** [comment]
Hello @ivarflakstad @LaurentMazare 

When can we expect @KimHenrikOtte 's PR (https://github.com/huggingface/candle/pull/3379) to be merged? Looking forward to having webgpu in Candle.
